buildSrc/src/main/kotlin/scripts/variants.gradle.kts

```diff
@@ -40,10 +40,23 @@ object BuildTypes {
 }
 
 object Default {
-    val BUILD_FLAVOR: String = System.getenv("flavor") ?: System.getenv("FLAVOR") ?: ProductFlavors.Dev.buildName
-    val BUILD_TYPE = System.getenv("buildType") ?: System.getenv("BUILD_TYPE") ?: BuildTypes.DEBUG
+    fun buildFlavour(default: String?): String? = System.getenv("flavor")
+        ?: System.getenv("FLAVOR")
+        ?: default
 
-    val BUILD_VARIANT = "${BUILD_FLAVOR.capitalize()}${BUILD_TYPE.capitalize()}"
+    fun buildFlavour(): String = System.getenv("flavor")
+        ?: System.getenv("FLAVOR")
+        ?: ProductFlavors.Dev.buildName
+
+    fun buildType(default: String?): String? = System.getenv("buildType")
+        ?: System.getenv("BUILD_TYPE")
+        ?: default
+
+    fun buildType(): String = System.getenv("buildType")
+        ?: System.getenv("BUILD_TYPE")
+        ?: BuildTypes.DEBUG
```

Comment:
I find weird to have functions with the same name, but one returns a concrete String and the other one returns an optional string.

I'd expect them to behave the same, and the only difference would be the optional parameter.

Also weird to have a class/object named `Default` that returns an `null` something when its scope would suggest it should return a "default" something.

Reply:
I'd maybe remove the `default: String?` parameter, as it is not helpful to the caller. It's a very minor sugar for an Elvis operator. And take it out of the `Default` object. Maybe move it to a `BuildIntention` object or similar, trying to figure out if there's any intention to build a specific app type.

----

```kotlin
buildFlavour("Default") == buildFlavour() ?: "Default
```

At least on the right side the caller has more control over it. On the left side you can't know if the result is "Default" because it actually is set to `"Default"`, or if it is because you passed the "Default" parameter.

Reply:
Alternatively:

```kotlin
object BuildIntention {
    fun explicitBuildFlavor(): String? = System.getenv("flavor")
        ?: System.getenv("FLAVOR")

    fun resolvedBuildFlavor(): String = explicitBuildFlavor() ?: ProductFlavors.Dev.buildName

    fun explicitBuildType(): String? = System.getenv("buildType")
        ?: System.getenv("BUILD_TYPE")

    fun resolvedBuildType(): String = explicitBuildType() ?: BuildTypes.DEBUG
}
```